internal/wire/frame_parser.go

```diff
@@ -82,114 +51,115 @@ func (p *FrameParser) parseNext(b []byte, encLevel protocol.EncryptionLevel, v p
-			frame, l, err = parseNewConnectionIDFrame(b, v)
-		case retireConnectionIDFrameType:
-			frame, l, err = parseRetireConnectionIDFrame(b, v)
-		case pathChallengeFrameType:
-			frame, l, err = parsePathChallengeFrame(b, v)
-		case pathResponseFrameType:
-			frame, l, err = parsePathResponseFrame(b, v)
-		case connectionCloseFrameType, applicationCloseFrameType:
-			frame, l, err = parseConnectionCloseFrame(b, typ, v)
-		case handshakeDoneFrameType:
-			frame = &HandshakeDoneFrame{}
-		case 0x30, 0x31:
-			if !p.supportsDatagrams {
-				return nil, 0, errUnknownFrameType
-			}
-			frame, l, err = parseDatagramFrame(b, typ, v)
-		case resetStreamAtFrameType:
-			if !p.supportsResetStreamAt {
-				return nil, 0, errUnknownFrameType
-			}
-			frame, l, err = parseResetStreamFrame(b, true, v)
-		default:
+	var err error
+	//nolint:exhaustive // Common frames should already be handled.
+	switch frameType {
+	case PingFrameType:
+		frame = &PingFrame{}
+		l = 0
+	case ResetStreamFrameType:
+		frame, l, err = parseResetStreamFrame(data, false, v)
+	case StopSendingFrameType:
+		frame, l, err = parseStopSendingFrame(data, v)
+	case CryptoFrameType:
+		frame, l, err = parseCryptoFrame(data, v)
+	case NewTokenFrameType:
+		frame, l, err = parseNewTokenFrame(data, v)
+	case MaxDataFrameType:
+		frame, l, err = parseMaxDataFrame(data, v)
+	case MaxStreamDataFrameType:
+		frame, l, err = parseMaxStreamDataFrame(data, v)
+	case BidiMaxStreamsFrameType, UniMaxStreamsFrameType:
+		frame, l, err = parseMaxStreamsFrame(data, frameType, v)
+	case DataBlockedFrameType:
+		frame, l, err = parseDataBlockedFrame(data, v)
+	case StreamDataBlockedFrameType:
+		frame, l, err = parseStreamDataBlockedFrame(data, v)
+	case BidiStreamBlockedFrameType, UniStreamBlockedFrameType:
+		frame, l, err = parseStreamsBlockedFrame(data, frameType, v)
+	case NewConnectionIDFrameType:
+		frame, l, err = parseNewConnectionIDFrame(data, v)
+	case RetireConnectionIDFrameType:
+		frame, l, err = parseRetireConnectionIDFrame(data, v)
+	case PathChallengeFrameType:
+		frame, l, err = parsePathChallengeFrame(data, v)
+	case PathResponseFrameType:
+		frame, l, err = parsePathResponseFrame(data, v)
+	case ConnectionCloseFrameType, ApplicationCloseFrameType:
+		frame, l, err = parseConnectionCloseFrame(data, frameType, v)
+	case HandshakeDoneFrameType:
+		frame = &HandshakeDoneFrame{}
+		l = 0
+	case ResetStreamAtFrameType:
+		if !p.supportsResetStreamAt {
 			err = errUnknownFrameType
+		} else {
+			frame, l, err = parseResetStreamFrame(data, true, v)
 		}
+	default:
+		err = errUnknownFrameType
 	}
 	if err != nil {
-		return nil, 0, err
-	}
-	if !p.isAllowedAtEncLevel(frame, encLevel) {
-		return nil, l, fmt.Errorf("%s not allowed at encryption level %s", reflect.TypeOf(frame).Elem().Name(), encLevel)
+		return frame, l, &qerr.TransportError{
+			ErrorCode:    qerr.FrameEncodingError,
+			FrameType:    uint64(frameType),
+			ErrorMessage: err.Error(),
+		}
 	}
-	return frame, l, nil
```

Comment:
You need to explicitly handle the error here. For example, parsing a STREAM frame will return `io.EOF` if the STREAM frame contains less data than the `Length` value indicates. This `io.EOF` must not be confused with reaching the end of the packet.

It should be sufficient to add a `if err != nil` here.